Add minecraft chart to incubator.

gtaylor opened the pull request:
Surprise!

gtaylor committed:
Add minecraft chart to stable.

viglesiasce commented:
@k8s-bot ok to test

prydonius commented on incubator/minecraft/README.md:
````diff
@@ -0,0 +1,64 @@
+
+```bash
+$ helm install --name my-release \
+    --set MinecraftServer.EULA=true incubator/minecraft
````
Can you check this works with Helm beta.2 given the following change:
https://github.com/kubernetes/helm/pull/1451

Given this will be interpreted as boolean, and the default is "FALSE" as a string.

gtaylor replied:
@prydonius I did this with the assumption that the `quote` filter here would stringify it:

https://github.com/kubernetes/charts/pull/154/files#diff-c198687fd5a49f25def79b8ce2933bdeR29

It seems to accept "TRUE" or "true" on the server side. Does the quote filter act like I think it does?

prydonius replied:
Just tested and it does indeed quote so it should be good, thanks!

prydonius commented:
Any chance in a NOTES.txt? Some short next steps for connecting to the Minecraft server (e.g. getting the IP address) and a warning if the EULA is set to false would be great.

gtaylor commented:
Sure. I just realized that NOTES.txt was a thing. That looks like a good idea.

chrislovecnm commented on incubator/minecraft/templates/datadir-pvc.yaml:
```diff
@@ -0,0 +1,14 @@
+metadata:
+  name: {{ template "fullname" . }}-datadir
+  annotations:
+    volume.alpha.kubernetes.io/storage-class: {{ .Values.Persistence.StorageClass | quote }}
```
Do you always need a PVC? How would I install this on minikube?

gtaylor replied:
The only scenario I can think of where you wouldn't want persistence is in testing. You can disable the PVC and use an emptyDir by setting `Values.Persistence.Datadir.Enabled` to false in the event that you don't.

chrislovecnm replied:
Perfect missed that

chrislovecnm commented on incubator/minecraft/templates/deployment.yaml:
```diff
@@ -0,0 +1,116 @@
+            cpu: "{{.Values.Cpu}}"
+            memory: "{{.Values.Memory}}"
+          limits:
+            cpu: "{{.Values.Cpu}}"
```
Is there a reason to tune limits and requests the same?

gtaylor replied:
I may be thinking about this wrong, but Minecraft tends to use what you give it. Aggressively. In this case, we've capped it at 756MB, so it's going to take what we give it as the process runs. I left some extra headroom in the numbers just in case.

I set the requests and limits to the same value since the actual usage should be very predictably just under a specific, user-configurable amount. It's kind of like Redis in how it'll chew up what you give it.

Open to other ideas if this isn't OK.

chrislovecnm replied:
So I would perfer to set both separately. You set mem in redis to the same value, because of the app. You don't want k8s sneaking mem from it... 

But I am not a minecraft guru.

gtaylor replied:
It was much the same when I last ran a Minecraft server (IE: I set this memory limit, it's going to use every bit of it like Redis under traffic).

Would it be OK to split these into separate values but leave the memory == limit, defined separately?

CPU request doesn't have to == limit, but Minecraft is similarly CPU hoggy. I could yank the limit and leave the request for a full CPU.

viglesiasce replied:
@gtaylor rather than having individual values for each you can make the entire dict a value as shown here:
https://github.com/kubernetes/charts/blob/master/stable/wordpress/values.yaml#L92

I prefer that method. Your defaults can stay the same as you have them now.

chrislovecnm replied:
Can we add that to the tech standards?

chrislovecnm commented on incubator/minecraft/templates/deployment.yaml:
```diff
@@ -0,0 +1,116 @@
+            memory: "{{.Values.Memory}}"
+        env:
+        - name: EULA
+          value: {{ .Values.MinecraftServer.EULA | quote }}
```
Holycrap I think you get a cookie for the most ENV vars on this repo. Thank goodness for charts.

chrislovecnm commented on incubator/minecraft/templates/secrets.yaml:
```diff
@@ -0,0 +1,12 @@
+    heritage: "{{ .Release.Service }}"
+type: Opaque
+data:
+  rcon-password: {{ default "" .Values.MinecraftServer.Rcon.Password | b64enc | quote }}
```
So secrets are not too secret. Known K8s challenge. Would a configmap be better?

gtaylor replied:
They're not too secret, but I'm not sure why I'd make it a configmap instead. It looks like Secrets are used in some of the other charts.

Hopefully the Secret situation improves as Kubernetes matures. That'd be what I'm banking on in using Secrets, as flimsy as they are now.

chrislovecnm replied:
Yah secrets have challenges. Sounds good to me

viglesiasce replied:
I agree with @gtaylor that although the implementation is not perfect we should not be banking on a feature that is not ever intended for keeping secrets. We hope that K8s will make this better over time, best to attach ourselves to the proper feature.

chrislovecnm replied:
Indeed

chrislovecnm commented on incubator/minecraft/values.yaml:
```diff
@@ -0,0 +1,82 @@
+  Rcon:
+    # If you enable this, make SURE to change your Password below.
+    Enabled: false
+    Port: 25575
```
Should we comment out the params that are off by default? Just a thought

gtaylor replied:
I took some time looking around at various charts trying to find a unifying style, but it seems like there's a lot of variation. This was one of the things I wasn't sure about.

I don't feel strongly either way. I guess I like the explicit values since I don't have to stick `default ""` gotmpl things everywhere, but that may not be a good enough reason.

chrislovecnm replied:
We need to add to the docs on a style we recommend. Appreciate you pointing that out.

viglesiasce replied:
Again agree with @gtaylor. I prefer an explicit value in the values.yml rather than implicit defaults at each point in the templates.

chrislovecnm reviewed (changes requested):
Thanks for the PR!!! Got some tweaks dos yah

gtaylor commented:
@chrislovecnm Thanks for the review. I'm not 100% sure which of the above are just questions/suggestions vs blockers for merging to incubator. Maybe that'll be more apparent after reading my responses.

As a sidenote, do we have any kind of style guide that I'm missing? I feel like I'm being really arbitrary with my capitalization and even some of the value organization. Some charts use a flat structure, others use something more nested like I am. Some use dropCase, others use CamelCase.

chrislovecnm commented:
@gtaylor looks great. I would ask that we break up request and resource limits.

We need to write recommended best practices / standards for this darn repo. There is an issue open. You brought up a bunch of great questions would love your feedback.

viglesiasce commented:
For best practices please send PRs here:
https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

If it is not yet listed there we shouldnt be blocking PRs from making it in.

chrislovecnm commented:
@viglesiasce agreed

gtaylor commented:
Some re-jiggering in fixup commits above. I've lowercased all of the values, since that's going to look better for using the nifty `toYaml` thing in the deployment (k8s manifests are lowercase). I've also added a NOTES.txt and adjusted the starting memory to 512 MB. This is workable for 1-3 players.

I can rebase+squash these prior to merge. Or a committer can, whatever is best.

chrislovecnm commented:
Not sure if I have perms here - @k8s-bot ok to test

Not sure what this repo prefers, but you squashing to chunks of work is always awesome.

chrislovecnm commented:
@k8s-bot ok to test

gtaylor commented:
Squashed. Now that we've gone through a few rounds of review, should I still be sending this to incubator? Or should I re-adjust this to be in stable? I think it meets all of the technical requirements now.

prydonius commented:
TLDR; lgtm!

My only concern in moving this to stable is the EULA requirement breaks the following requirement:

> Must successfully launch with default values (helm install .)
> All pods go to the running state
> All services have at least one endpoint

from [CONTRIBUTING.md](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md).

However, given the NOTES.txt appropriately warns about this, and it's reasonable to require a EULA, I'm okay with this going into stable.

Just wanted to point it out, since this could cause issues when the CI does more functional testing of Charts.

prydonius commented on incubator/minecraft/templates/NOTES.txt:
```diff
@@ -0,0 +1,46 @@
+Getting Started:
```
Can you remove this line? It's redundant since the header helm prints out is `NOTES:`.

viglesiasce commented:
You could make an if block around the deployment when the EULA is set to false, then you wouldnt deploy it at all and the NOTES could further note that you dont even have it deployed yet.

prydonius commented:
> You could make an if block around the deployment when the EULA is set to false, then you wouldnt deploy it at all and the NOTES could further note that you dont even have it deployed yet.

Actually this is an interesting idea. Then, the command to add the deployment could simply be `helm upgrade <release> --set minecraftServer.eula=true`. Actually, this command should also work for the current case too.

prydonius commented on incubator/minecraft/templates/NOTES.txt:
```diff
@@ -0,0 +1,46 @@
+Getting Started:
+
+{{- if eq (.Values.minecraftServer.eula | quote) "FALSE" }}
```
This didn't work for me, it should just be

```
{{- if eq .Values.minecraftServer.eula "FALSE" }}
```

prydonius replied:
Ah the issue is this doesn't work when you use `--set` because it doesn't work on booleans.

prydonius replied:
```
{{- if eq (printf "%s" .Values.minecraftServer.eula) "FALSE" }}
```

should work for both cases

gtaylor commented:
@prydonius I think I got everything addressed in the fixup commits.

prydonius commented on stable/minecraft/README.md:
```diff
@@ -0,0 +1,65 @@
+
+## Configuration
+
+Refer to `values.yaml` for the full run-down on defaults. These are a mixture of Kubernetes and Minecraft-related directives that map to environment variables in the [itzg/minecraft-server](https://hub.docker.com/r/itzg/minecraft-server/) Docker image.
```
It would be good to add a link to the values.yaml - https://github.com/kubernetes/charts/blob/master/stable/minecraft/values.yaml

prydonius commented on stable/minecraft/templates/NOTES.txt:
```diff
@@ -1,10 +1,18 @@
+
+This deployment will be incomplete and need to be deleted:
+
+    helm delete {{ .Release.Name }}
```
LGTM as is, but a quicker way, is to keep the release and add the deployment by performing an upgrade:

```
helm upgrade {{ .Release.Name }} \
  --set minecraftServer.eula=true stable/minecraft
```

works nicely.

Up to you, one potential problem could be that I'm not 100% sure that upgrading a release to the same chart version is actually supported. cc/ @michelleN

viglesiasce commented:
I've been doing upgrades to the same version over the last few days and havent hit any issues.

gtaylor commented:
@prydonius Added the link to README.md and updated NOTES.txt as suggested.

prydonius commented:
@gtaylor lgtm, great work :)!
